### PR TITLE
connected assisted installer

### PR DESCRIPTION
**Pull request by dhellmann**
This enhancement describes changes in and around the installer to
assist with deployment on user-provisioned infrastructure. The use
cases are primarily relevant for bare metal, but in the future may be
applicable to cloud users who are running an installer UI directly
instead of using a front-end such as `cloud.redhat.com` or the UI
provided by their cloud vendor.

### CODE REVIEW

**Comment by bparees** [comment]
@dhellmann not clear why i'm an assignee on this, is there a particular aspect you're looking for my feedback on?

**Comment by dhellmann** [comment]
> @dhellmann not clear why i'm an assignee on this, is there a particular aspect you're looking for my feedback on?

The bot gave me your name. I assumed it had some way to pick you as a reviewer other than /dev/random but perhaps not?

**Comment by bparees** [comment]
pretty much randomly selected from here: https://github.com/openshift/enhancements/blob/master/OWNERS

@sdodson probably more your area, or maybe you can pick someone more appropriate to be involved in reviewing this

/unassign @bparees

**Comment by dhellmann** [comment]
OK, thanks @bparees 

/assign @sdodson

**Comment by cgwalters** [comment]
I only skimmed so far, but I love the ideas behind this!

This is building extensively on https://github.com/openshift/enhancements/pull/210 and it'd be good to cross-link.

File: enhancements/installer/connected-assisted-installer.md [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+cases are primarily relevant for bare metal, but in the future may be
+applicable to cloud users who are running an installer UI directly
+instead of using a front-end such as `cloud.redhat.com` or the UI
+provided by their cloud vendor.
+
+## Motivation
+
+The target user is someone wanting to deploy OpenShift, especially on
+bare metal, with as little up-front infrastructure dependencies as
+possible. This person has access to servers and wants to run workloads
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:46]: "to servers" is not very specific.  You mean "to cloud.redhat.com", right?

File: enhancements/installer/connected-assisted-installer.md [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+quickly. They do not necessarily have the administrative privileges to
+create private VLANs, configure DHCP/PXE servers, or manage other
+aspects of the infrastructure surrounding the hardware where the
+cluster will run. If they do have the required privileges, they may
+not want to delegate them to the OpenShift installer for an IPI
+installation, preferring instead to use their existing tools and
+processes for some or all of that configuration. They are willing to
+accept that the cluster they build may not have all of the same
+features as an IPI cluster on day 1, but that by taking additional
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:55]: nit: "an IPI cluster" -> "a cluster with installer-provisioned infrastructure"

File: enhancements/installer/connected-assisted-installer.md (hunk 1 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md:46]: No, here we mean that they have physical server hardware of their own on which they want to run OCP to manage workloads.

File: enhancements/installer/connected-assisted-installer.md [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+steps later they will be able to add more automatic infrastructure
+management and make it fully production-ready.
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:57]: User-provisioned infrastructure doesn't mean a cluster isn't production-ready.  Maybe just stop with "automatic infrastructure management."?

File: enhancements/installer/connected-assisted-installer.md (hunk 1 above)
Comment (mhrivnak) [reply enhancements/installer/connected-assisted-installer.md:46]: I think this means "has access to server-grade hardware on which they want to install OpenShift and then run workloads quickly".

File: enhancements/installer/connected-assisted-installer.md [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+### Goals
+
+- Make initial deployment of usable and supportable clusters simpler.
+- Move more infrastructure configuration from day 1 to day 2.
+- Support connected on-premise deployments.
+- Support existing IPI features, especially for day 2 cluster
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:64]: nit: provisioner shouldn't matter for features, especially day-2 features.  This should be:

> * Support cluster-managed infrastructure features, especially on day 2.

Or something like that...

File: enhancements/installer/connected-assisted-installer.md [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+### Non-Goals
+
+- Because the initial focus is on bare metal, this enhancement does
+  not exhaustively cover variations needed to offer similar features
+  on other platforms (such as changes to image formats, the way a host
+  boots, etc.). It is desirable to support those platforms, but that
+  work will be described separately.
+- Environments where hosts cannot reach the internet unimpeded
+  ("disconnected" or "air-gapped") will require more work to support
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:75]: nit: Canonical wording for this is "[restricted network][1]", so you should include that in your list of synonyms.

[1]: https://github.com/openshift/openshift-docs/blob/50e2b45e6c90f6956b916e11133fb83cb08e4e6e/modules/installation-about-restricted-network.adoc

File: enhancements/installer/connected-assisted-installer.md (hunk 1 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md]: I added the word "hardware" to the sentence to clarify.

File: enhancements/installer/connected-assisted-installer.md [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+
+The user then boots the live ISO on each host they want to be part of
+the cluster (control plane and workers). They can do this by hand
+using thumb drives, by attaching the ISO using virtual media support
+in the BMC of the host, or any other way they choose.
+
+When the ISO boots, it starts an agent that communicates with the REST
+API for the assisted installer service running on `cloud.redhat.com`
+to receive instructions. The agent registers the host with the
+service, using the user's pull secret embedded in the ISO's ignition
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:104]: nit: "ignition" -> "Ignition config" to match our [canonical wording][1].

[1]: https://github.com/openshift/openshift-docs/blame/88033b5b246b6510ce89e5165f1fb9885f9a7a47/modules/rhcos-about.adoc#L24

File: enhancements/installer/connected-assisted-installer.md (hunk 2 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md]: fixed

File: enhancements/installer/connected-assisted-installer.md (hunk 3 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md:58]: This has more to do with waiting to take some setup steps until day 2, or not doing them at all, than it does with UPI.

File: enhancements/installer/connected-assisted-installer.md (hunk 4 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md:65]: I'm not aware that it is currently possible to install a UPI bare metal cluster and then on day 2 use IPI to add workers. That's the goal here.

File: enhancements/installer/connected-assisted-installer.md (hunk 5 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md:76]: done

File: enhancements/installer/connected-assisted-installer.md (hunk 6 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md]: done

File: enhancements/installer/connected-assisted-installer.md (hunk 4 above)
Comment (mhrivnak) [reply enhancements/installer/connected-assisted-installer.md:65]: There are features that you get today on bare metal only if you used IPI and do not get with UPI, particularly power management via BMC. Given the direction we're going, it probably makes sense for us to stop referring to that as "existing IPI features" and get in the habit of using a more direct descriptor.

Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md:65]: Are we really just disagreeing about the wording? Or is the description unclear?

File: enhancements/installer/connected-assisted-installer.md [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+machine-config-operator in the usual way.
+
+During installation, progress and error information is reported to the
+assisted installer application on `cloud.redhat.com` so it can be
+shown in the UI.
+
+### Integration with Existing Bare Metal IPI Tools
+
+Clusters built using the assisted installer workflow use the same
+"baremetal" platform setting as clusters built with the IPI
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:155]: More IPI references here and in the subsection header.  Maybe:

> ### Integration with Existing IBare Metal Tools for Installer-Provisioned Infrastructure
>
> Clusters built using the assisted installer workflow use the same "baremetal" platform setting as clusters with installer-provisioned infrastructure.  The cluster runs...

File: enhancements/installer/connected-assisted-installer.md [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+of hosts using my existing provisioning tools to automate launching
+the installer so I can adapt my existing admin processes and
+infrastructure tools instead of replacing them.
+
+#### Story 3
+
+As a cluster deployer, I want to install a production-ready OpenShift
+cluster without committing to delegating all infrastructure control to
+the installer or to the cluster, so I can adapt my existing admin
+processes and infrastructure management tools instead of replacing
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:184]: This story seems orthogonal to the enhancement.  Folks can already do this with user-provisioned infrastructure, and afterwards delegate as much or as little of the infrastructure management as they want to the cluster, right?  If there are missing delegation bits, sorting those out seems like it would be a day-2 issue, and this enhancement is about day-1 issues.  Or am I missing a connection...?

File: enhancements/installer/connected-assisted-installer.md [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+The user then boots the live ISO on each host they want to be part of
+the cluster (control plane and workers). They can do this by hand
+using thumb drives, by attaching the ISO using virtual media support
+in the BMC of the host, or any other way they choose.
+
+When the ISO boots, it starts an agent that communicates with the REST
+API for the assisted installer service running on `cloud.redhat.com`
+to receive instructions. The agent registers the host with the
+service, using the user's pull secret embedded in the ISO's ignition config
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:105]: nit: "ignition" -> "Ignition", per [this thread][1].

[1]: https://github.com/openshift/enhancements/pull/376#discussion_r440459336

File: enhancements/installer/connected-assisted-installer.md (hunk 4 above)
Comment (wking) [reply enhancements/installer/connected-assisted-installer.md:65]: I had to look up [baseboard management controller][1], so take this with a grain of "Trevor has no idea what he's talking about" salt ;).  But it seems to me that "I have an existing metal cluster and want to give its machine API controller the ability and authority to manage Machines as a day 2 operator" would be a different enhancement from (and possibly a prerequisite for) this install-time enhancement.  Couldn't you set up that day-2 management without needing to call out to `cloud.redhat.com`?  You'd already have a local cluster where you could host whatever the incoming machines would need for their first boot...

[1]: https://en.wikipedia.org/wiki/Intelligent_Platform_Management_Interface#Baseboard_management_controller

File: enhancements/installer/connected-assisted-installer.md [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+need to be moved into more a suitable GitHub org.
+
+The agent discussed in this design is different from the
+`ironic-python-agent` used by Ironic in the current IPI
+implementation.
+
+### Risks and Mitigations
+
+The current implementation relies on
+[minimise-baremetal-footprint](../baremetal/minimise-baremetal-footprint.md). If
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:207]: A reference to the in-flight #361?  What if this enhancement lands but that one does not?  Maybe link the PR instead?

File: enhancements/installer/connected-assisted-installer.md [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+that approach cannot be supported, users can proceed by providing an
+extra host (4 hosts to build a 3 node cluster, 6 hosts to build a 5
+node cluster, etc.).
+
+## Design Details
+
+### Open Questions
+
+N/A
```
Comment (wking) [inline enhancements/installer/connected-assisted-installer.md:216]: nit: you can probably drop the whole _Open Questions_ subsection if you have no open questions.

File: enhancements/installer/connected-assisted-installer.md (hunk 3 above)
Comment (wking) [reply enhancements/installer/connected-assisted-installer.md:58]: Then probably drop both "installer-provisioned infrastructure" and "fully production-ready" and just focused on "day 2 changes to automate infrastructure management"?

**Comment by ericavonb** [comment]
Could you add some specifics notes on the security for this design? There are a lot of attack vectors in remote installation and auto-join systems, but maybe there are details I’m missing that make this model secure. Would be helpful to have them explicitly noted.

File: enhancements/installer/connected-assisted-installer.md (hunk 9 above)
Comment (eranco74) [inline enhancements/installer/connected-assisted-installer.md:105]: @romfreiman do we plan to embed the user pull secret to the ISO?
Currently this isn't the case and I think we can use some other token (that the cloud service generate) for the authentication rather than the user pull secret. 
In general I think it's preferable not to have user secrets / passwords in the ISO.

File: enhancements/installer/connected-assisted-installer.md [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+
+The discovered hardware and networking details are combined with the
+results of the validation to derive defaults for the machine network
+CIDR, the API VIP, and other network configuration settings for the
+hosts.
+
+When enough hosts are configured, the assisted installer application
+replies to the agent on each host with the instructions it needs to
+take part in forming the cluster. The assisted installer application
+selects one host to run the bootstrap services used during
```
Comment (eranco74) [inline enhancements/installer/connected-assisted-installer.md:143]: In regard to the bootstrap service.
I think it's worth mentioning that unlike all other OpenShift installers, the assisted installer doesn’t use an auxiliary bootstrap node.
Instead, it will pivot the bootstrap to become a master node once the control plane is running on 2 other master nodes.
This flow reduces the minimum number of nodes to 3.

Comment (wking) [reply enhancements/installer/connected-assisted-installer.md:143]: > Instead, it will pivot the bootstrap to become a master node once the control plane is running on 2 other master nodes.

Do we run a descheduler by default now?  In addition to the "did we clean out all the bootstrap-specific stuff?" concerns we had about pivoting bootstrap into a compute node, pivoting into a control-plane node adds "do we end up with most of the critical pieces all lumped together on the two born-as-control-plane machines?".  Although I guess a few rounds of control-plane reboots during subsequent updates and config rollouts would wash away any initial lopsided balancing.

File: enhancements/installer/connected-assisted-installer.md (hunk 9 above)
Comment (wking) [reply enhancements/installer/connected-assisted-installer.md:105]: If we have concerns about holding user secrets server-side, this could be the service encrypting the pull secret to itself, storing the encrypted blob in the ISO, and just hanging on to the encryption key.

File: enhancements/installer/connected-assisted-installer.md (hunk 12 above)
Comment (cgwalters) [reply enhancements/installer/connected-assisted-installer.md:143]: >  In addition to the "did we clean out all the bootstrap-specific stuff?" concerns we had about pivoting bootstrap into a compute node, 

This is a pretty easy problem to solve, we can add support to Ignition to make everything ephemeral at the OS level (i.e. mount `/var` as a `tmpfs`, `/etc` as an overlayfs and everything else read-only).  Then rebooting guarantees everything done at the filesystem level is gone (and there's no reason for an OpenShift install process to operate at the block level).

File: enhancements/installer/connected-assisted-installer.md (hunk 4 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md]: You're right, the two things are somewhat related.

The existing bare metal IPI workflow requires management controller credentials for each host, so the installer can automate the deployment. Under that flow the installer has all of the information that would be required to turn the machine API on, so it does not make much sense to assume the machine API is disabled and we expect to always leave it enabled.

It also should be possible to use the UPI process to install a cluster and then add the machine API integration later. There's a separate set of work planned for that, regardless of the progress on this enhancement. For a UPI process, the user is going to have to do more work to set up the machine API, including BareMetalHost CRs with specific settings and connecting them to Machine CRs.

The important thing we're referring to here is that we can save some of the information available during the new automated installation process so that the user has less work to do to enable the machine API later. It seemed like adding the details of how that would work would be going too deep into planned implementation, but if we need more detail in this document about how that would work we can add it.

File: enhancements/installer/connected-assisted-installer.md (hunk 12 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md:142]: This part of the discussion should probably move to #361.

File: enhancements/installer/connected-assisted-installer.md (hunk 8 above)
Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md:184]: We don't currently make it easy for them to add the bare metal machine API on day 2 (see discussion above). Being able to do that is a key requirement for this work. Perhaps that deserves its own enhancement?

**Comment by dhellmann** [comment]
> Could you add some specifics notes on the security for this design? There are a lot of attack vectors in remote installation and auto-join systems, but maybe there are details I’m missing that make this model secure. Would be helpful to have them explicitly noted.

@avishayt @romfreiman maybe we can work on responding to this request together?

**Comment by avishayt** [comment]
> > Could you add some specifics notes on the security for this design? There are a lot of attack vectors in remote installation and auto-join systems, but maybe there are details I’m missing that make this model secure. Would be helpful to have them explicitly noted.
> 
> @avishayt @romfreiman maybe we can work on responding to this request together?

We are beginning to formalize this and will provide the information once we have it compiled.

File: enhancements/installer/connected-assisted-installer.md (hunk 4 above)
Comment (hardys) [reply enhancements/installer/connected-assisted-installer.md]: The current interface which decides if cluster-managed-baremetal pieces are enabled is the platform, so when you do an IPI deployment using `platform: baremetal` you get those features.

However the current UPI docs use `platform: none` and there's no currently supported way to enable the managed-baremetal pieces after initial deployment.  There has been some work looking at how we might change that ref https://github.com/openshift/machine-api-operator/pull/560

However I don't think making day-2 enablement of managed baremetal is a prerequisite for this assisted-installer workflow, because AFAIK the plan is to use the `platform: baremetal` even in the case where you choose the "boot it yourself" path - this also gives some other features like cluster-hosted loadbalancing and DNS, so it makes sense to reuse the same platform if possible.  /cc @avishat to keep me honest on plans here :)

Comment (romfreiman) [reply enhancements/installer/connected-assisted-installer.md]: Indeed, we use installer platform: baremetal to generate the ignitions

File: enhancements/installer/connected-assisted-installer.md [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+* https://github.com/eranco74/assisted-installer.git : The actual
+  installer code that runs on the hosts
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
+
+The telco/edge bare metal team is working on support for automating
```
Comment (markmc) [inline enhancements/installer/connected-assisted-installer.md:310]:
```suggestion
The bare metal IPI team is working on support for automating
```

**Comment by markmc** [comment]
> > > Could you add some specifics notes on the security for this design? There are a lot of attack vectors in remote installation and auto-join systems, but maybe there are details I’m missing that make this model secure. Would be helpful to have them explicitly noted.
> > 
> > 
> > @avishayt @romfreiman maybe we can work on responding to this request together?
> 
> We are beginning to formalize this and will provide the information once we have it compiled.

Definitely think @ericavonb is spot on with this request. Perhaps it can be tracked as a TODO or graduation requirement if y'all want to move forward with merging what has been written so far (and seems largely agreed on)

/lgtm

File: enhancements/installer/connected-assisted-installer.md (hunk 13 above)
Comment (wking) [reply enhancements/installer/connected-assisted-installer.md:310]: We cannot say "IPI" in external docs.  We need to say "installer-provisioned infrastructure".  But do we really have a metal team that ignores user-provisioned infrastructure?

Comment (dhellmann) [reply enhancements/installer/connected-assisted-installer.md:310]: Fixed in #382 

The team working on the automation does not support user-provisioned deployments today. IIUC, the main installer team supports those.
